### PR TITLE
Add support for Razer Ornata V3

### LINKED ISSUE #1925: Support for Razer Ornata V3 [028F]
### Device Info

* Name: Razer Ornata v3
* Product Number: RZ03-04460300-R3W1
* Model Number: RZ03-0446

### Device Information
```
Please post the output of
    lsusb -d 1532: -v

Bus 001 Device 029: ID 1532:028f Razer USA, Ltd Razer Ornata V3
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x028f 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Ornata V3
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1

```

```
Please post the output of
    ls -lah /dev/input/by-id/

lrwxrwxrwx 1 root root  10 Oct 11 16:39 usb-Razer_Razer_Ornata_V3-event-if01 -> ../event20
lrwxrwxrwx 1 root root   9 Oct 11 16:39 usb-Razer_Razer_Ornata_V3-event-kbd -> ../event2
lrwxrwxrwx 1 root root   9 Oct 11 16:39 usb-Razer_Razer_Ornata_V3-if01-event-kbd -> ../event4
lrwxrwxrwx 1 root root  10 Oct 11 16:39 usb-Razer_Razer_Ornata_V3-if02-event-mouse -> ../event21
lrwxrwxrwx 1 root root   9 Oct 11 16:39 usb-Razer_Razer_Ornata_V3-if02-mouse -> ../mouse0

```

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on **Razer Synapse** in a Windows Environment>

### PR DESCRIPTION
**Pull request by ddB0515**
This PR will add support for Ornata V3 and fixes #1925 
Tested and all working

### CODE REVIEW

**Comment by ddB0515** [comment]
Not sure why job failed as tried to run script on my PC but gets weird error
```
$ ./scripts/format_source.sh
+ find driver/ -name '*.c' -o -name '*.h'
+ xargs astyle --style=linux -n
xargs: astyle: No such file or directory
```

Not sure if that is a problem as I don't know how to solve it :(

**Comment by z3ntu** [comment]
You need to install `astyle` package for formatting C source. `xargs` is just a tool to run the command on each of the files

**Comment by ddB0515** [comment]
@z3ntu now I have build fresh ones and on Spectrum animation when I type lights goes OFF and after few sec they turn ON again 
I know that is not normal but not sure what can cause it :thinking: and how to fix it

**Comment by z3ntu** [comment]
Not sure what's happening there.. sorry

File: README.md [hunk 1]
```diff
@@ -75,6 +75,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Blade Stealth (Late 2020)               |  1532:0259  |
 | Razer BlackWidow V3 Pro (Wired)               |  1532:025A  |
 | Razer Ornata V2                               |  1532:025D  |
+| Razer Ornata V3                               |  1532:028F  |
```
Comment (z3ntu) [inline README.md]: Make sure you keep this list sorted by USB PID (so put it after `1532:028B`)

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -894,6 +896,9 @@ static ssize_t razer_attr_read_device_type(struct device *dev, struct device_att
         device_type = "Razer Ornata V2\n";
         break;
 
+    case USB_DEVICE_ID_RAZER_ORNATA_V3:
+        device_type = "Razer Ornata V3\n";
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:985]: newline after this

File: driver/razerkbd_driver.h [hunk 3]
```diff
@@ -72,6 +72,7 @@
 #define USB_DEVICE_ID_RAZER_BLADE_STEALTH_LATE_2020 0x0259
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_WIRED 0x025A
 #define USB_DEVICE_ID_RAZER_ORNATA_V2 0x025D
+#define USB_DEVICE_ID_RAZER_ORNATA_V3 0x028F
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: keep this list sorted by value also

**Comment by z3ntu** [comment]
Diff looks sane apart from the few comments I've added

**Comment by TECH7Fox** [comment]
I have the same keyboard if you need some testing.

**Comment by ddB0515** [comment]
Hi @TECH7Fox 
If you can checkout my branch and install/test, should work but for some reason after all changes  on Spectrum animation when I type lights goes OFF and after few sec they turn ON
If this is happening to you also please let me know will try to investigate/fix it
Thanks

**Comment by TECH7Fox** [comment]
Hey @ddB0515,

I have the same issue. With static it just stays off on 1 keypress. very weird...

Also, not sure if this is only with this keyboard, the wave effect seems to be inverted. (left = right, right = left)
I did notice some flicker when typing with the wave effect, once every minute it seems like. And the ripple effect is has a 0,5 sec delay, while reactive seems pretty instant.

Again, first time ever using openrazer so not sure if this is a keyboard specific issue. I use polychromatic as GUI.

Might be related, but when I hold c and d, and press spacebar I get `cdg `. It adds a g. Also when openrazer is off. Do you also have that?

Will let you know if I encounter anything else.

**Comment by TECH7Fox** [comment]
Just noticed that it works in regions for some effects, including reactive. In the effect editor it's a 1x9 grid. Is this a hardware thing?

**Comment by nightsky30** [comment]
@ddB0515 Although this is a very different type of device, some of this sounds familiar to behavior noticed with the Charging Pad Chroma or Thunderbolt 4 Dock Chroma.

Both of their wave directions were reversed as @TECH7Fox mentioned noticing with this Ornata V3. You may want to try something like this:

https://github.com/openrazer/openrazer/blob/8f50fda0be5c98b545d9c04ddef1131d37732add/driver/razeraccessory_driver.c#L629-L638
https://github.com/openrazer/openrazer/blob/8f50fda0be5c98b545d9c04ddef1131d37732add/driver/razeraccessory_driver.c#L1507-L1513

Also similar, the Charging Pad Chroma displayed some strange effect behavior if the device wasn't set to the proper mode. You might try explicitly switching between driver and normal device modes, and see if the lighting effects stay active during key presses:

https://github.com/openrazer/openrazer/blob/8f50fda0be5c98b545d9c04ddef1131d37732add/driver/razeraccessory_driver.c#L629-L631
https://github.com/openrazer/openrazer/blob/8f50fda0be5c98b545d9c04ddef1131d37732add/driver/razeraccessory_driver.c#L891-L896

**Comment by ddB0515** [comment]
Thanks @nightsky30 and @TECH7Fox for info
I tough will be easy support but looks like I was wrong :(  anyway will take a look next week as a bit busy with life currently

**Comment by z3ntu** [comment]
Were you able to fix the issues that were talked about above?

**Comment by lukester1975** [comment]
I've been playing around with this. 

Some good news: I have a workaround for the wacky static behaviour. If you create a custom effect in polychromatic with all segments set to the same colour, it mostly works.

Problem is the V3 appears to be a 1 * 10 matrix, not 1 * 9. I've put some fixes here: https://github.com/lukester1975/openrazer/tree/ornata-v3-hackery

That also has a bunch of other changes from comparing usb captures from Windows. Bad news is nothing yet I've tried "fixes" the static mode (although the problem of going blank after a keypress affects other effects, too). And it may well not be possible. Looking at the Windows capture, the request sent is:

```
0000   00 1f 00 00 00 23 0f 03|00 00 00 00 09|11 22 33
0010   11 22 33 11 22 33 11 22 33 11 22 33 11 22 33 11
0020   22 33 11 22 33 11 22 33 11 22 33|00 00 00 00 00
0030   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0040   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0050   00 00 00 00 00 00 00 00|26 00

transaction_id = 0x1f
data_size = 0x23
command_class = 0x0f
command_id = 0x03
arguments = 0x00 0x00 0x00 0x09 
   0x11 0x22 0x33 repeat 10 in total
```

So that's using the custom effect too, though does it in one shot with a single message. I may delve into it a bit further eventually but happy for now to have something working!

Can separate out the non-Ornata fixes to PRs if preferred.

HTH

**Comment by ddB0515** [comment]
Appologies not been active due to work/life, back and have some free time so if anyone is willing to try to get this back on track and do it
@lukester1975 did you got anywhere after doing latest changes? Any news

Also need to update main branch to current to get all up to speed...

**Comment by ddB0515** [comment]
I have fix some of issues and tested works fine (not tested a lot tbh)
* Wave have proper direction and works fine as suggested by @nightsky30 in top comments
* Changed some types to match findings from @lukester1975 
* Removed unused modes (that does not exist for this type of keyboard)

Still have some issue with build failing due to 
```
FAILED: Razer Ornata V3
        embedded null byte

FAILED: Razer Ornata V3
        0x28f Has capability 'lighting_led_matrix' but no sysfs file: 'matrix_custom_frame'

FAILED: Razer Ornata V3
        0x28f Has capability 'lighting_led_matrix' but no sysfs file: 'matrix_effect_custom'
```
But not sure what that is for (or why is failing...need to find out (maybe @z3ntu can help? )

If anyone can pull, build and test would appreciate it
Thanks

**Comment by ddB0515** [comment]
So I have managed to finally get all decoded from pcapng and this is transition from STATIC to SPECTRUM

```
Static
00 1F 000000 23 0F 03 00 00 0000 0900 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000
00 1F 000000 23 0F 03 00 00 0000 0900 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000
...
Spectrum
...
00 1F 000000 23 0F 03 00 00 0000 0900 FB00 00FB 0000 FB00 00FB 0000 FB00 00FB 0000 FB00 00FB 0000 FB00 00FB 0000
00 1F 000000 23 0F 03 00 00 0000 0903 F503 03F5 0303 F503 03F5 0303 F503 03F5 0303 F503 03F5 0303 F503 03F5 0300
00 1F 000000 23 0F 03 00 00 0000 0906 EF06 06EF 0606 EF06 06EF 0606 EF06 06EF 0606 EF06 06EF 0606 EF06 06EF 0600
...(a lot in between so didn't paste all but will attach whole file)
00 1F 000000 23 0F 03 00 00 0000 0901 F100 01F1 0001 F100 01F1 0001 F100 01F1 0001 F100 01F1 0001 F100 01F1 
00 1F 000000 23 0F 03 00 00 0000 0901 F500 01F5 0001 F500 01F5 0001 F500 01F5 0001 F500 01F5 0001 F500 01F5
00 1F 000000 23 0F 03 00 00 0000 0901 F800 01F8 0001 F800 01F8 0001 F800 01F8 0001 F800 01F8 0001 F800 01F8
...
Static again
...
00 1F 000000 23 0F 03 00 00 0000 0900 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000
00 1F 000000 23 0F 03 00 00 0000 0900 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000 FF00 00FF 0000
```

As @lukester1975 confirmed I got same, PR will need a bit of love to get this whole setup working...
Will keep you posted but if someone got anything working please let me know...Thanks

Download whole log in attachent
[STATIC2SPCTRUM.PCAPNG.log](https://github.com/openrazer/openrazer/files/11083913/STATIC2SPCTRUM.PCAPNG.log)

**Comment by michimussato** [comment]
Hope this gets through soon. Thanks everyone for all the work to this point!

**Comment by rmysliwczyk** [comment]
I'm not sure if this is helpful, but I was able to build this and it works on my Ornata V3. However, when I use polychromatic to select the STATIC effect I've noticed a problem: the lighting effect turns off after any key is pressed.

**Comment by ddB0515** [comment]
@rmysliwczyk  thanks for reporting, does everything else works fine in polychromatic?
Can you tell me steps how I can reproduce so I can take a look?

**Comment by rmysliwczyk** [comment]
> @rmysliwczyk thanks for reporting, does everything else works fine in polychromatic? Can you tell me steps how I can reproduce so I can take a look?

I'm sorry but I had to return my keyboard because it had a faulty key, so I can't do any further testing. What I did was I simply selected the static lighting effect from polychromatic, which applied the effect, but when I went to press any of the keys after doing that the lights would go off. They would go back on only after changing the effect to a different one and back.

**Comment by z3ntu** [comment]
I think the current CI failure ("embedded null byte") is caused by `HAS_MATRIX = True` but not having added the method `lighting_led_matrix` so you probably have to add that to the device.

**Comment by ddB0515** [comment]
Thanks @z3ntu 
Now is building with no errors hope now more people will jump to test this and to finally get it ready for merge

**Comment by z3ntu** [comment]
@ddB0515 What's the latest state of functionality for this device? Everything works as expected or still some issues?

**Comment by ddB0515** [comment]
Hello @z3ntu 
There is one more issue, I have uploaded video so check it out (not sure if that is something easy to fix but I don't know what else I can do about it :( (https://photos.app.goo.gl/mUSLHuTRMqS8Ge6R6)

If you have any suggestion let me know

**Comment by z3ntu** [comment]
Hm that's a very weird issue. I know in `razermouse_driver.c` some devices also have weird workarounds for basically misbehaving firmware. Maybe you can find something there?

**Comment by ddB0515** [comment]
Will take a look for sure also one option is to remove that static one from modes  :) (not sure if that's going to be an issue?)

**Comment by z3ntu** [comment]
Yep, worst case just remove static option, while it's not perfect (obviously) it's good enough of a solution to get the device included into OpenRazer if you can't figure out a fix for that problem.

**Comment by ddB0515** [comment]
Will wait if someone else can verify that is same for them (on other keyboards) that is not maybe my keyboard wrong or defective
If we get this confirmed that is same will remove static and we can merge it (with maybe fix/adding it back later) 
Will keep you posted as user from Telegram got to me that is willing to try

**Comment by ddB0515** [comment]
Anyone here have keyboard that is willing to test updated PR??? And help me resolve issue with static option?

**Comment by da-kurlzzzzz** [comment]
> Anyone here have keyboard that is willing to test updated PR??? And help me resolve issue with static option?

@ddB0515, I have an Ornata v3. How can I help? I am new to the field of keyboards so there might be lots of stupid questions. 

I have Arch and Win10 installed on my PC if that's relevant to possible testing

**Comment by ddB0515** [comment]
Excellent @da-kurlzzzzz 
Would be nice to move to Telegram group/chat as would be easier, add me there so we can continue and after test will update here progress

**Comment by michimussato** [comment]
Hi Dario

I own an Ornata V3 too running Manjaro. Could help out too.

Cheers,
Michael

**Comment by ddB0515** [comment]
@michimussato that would be lovely 
as last night we discovered that @da-kurlzzzzz have same keyboard (different  revision) and different PID
Can you please check before we started what device PID you have? 
Run: `lsusb -d 1532: -v` and check `idProduct` if is same as mine 028F or something else?

@z3ntu 
How is usually handled if device have different PID?  As new device or you never had that before?

**Comment by michimussato** [comment]
lsusb -d 1532: -v | grep -i idProduct
Couldn't open device, some information will be missing
  idProduct          0x028f

Hope that helps!

Am Di., 20. Juni 2023 um 10:25 Uhr schrieb Dario Budimir <
***@***.***>:

**Comment by ddB0515** [comment]
@michimussato that's super nice as is same as mine :muscle: 
Now you will need to do checkout of my branch and compile/install it on your device so you can do test

**Comment by michimussato** [comment]
@ddB0515 check out https://github.com/ddB0515/openrazer/tree/ornata-v3-support or https://github.com/ddB0515/openrazer/tree/ornata-v3-r1-support?

**Comment by ddB0515** [comment]
This one: https://github.com/ddB0515/openrazer/tree/ornata-v3-support 
`r1` is specific for test only (regarding new rev that I mention it before)

Also connect to Telegram: https://t.me/ddB0515 so we can continue faster communication if is not a problem

**Comment by armakunizen** [comment]
Does someone know if the Razer Ornata V3 and the Razer Ornata V3 X are similar devices?

`lsusb -d 1532:` gives:

`Bus 003 Device 006: ID 1532:02a2 Razer USA, Ltd Razer Ornata V3 X
`

**Comment by ddB0515** [comment]
Well same issue we just found out...looks like Razer have same produces with different PID
So what you can do is take [PR for existing Ornata V3 X](https://github.com/openrazer/openrazer/pull/2052) and modify to yours PID, test and make PR for it (as `R1` or whatever difference is)

@z3ntu looks like we are getting more and more same models different PID :)

**Comment by erikzmmrmnn** [comment]
I have an Ornata V3 with product ID 028f. I just tested setting the static lighting mode and it works perfectly for me. No turning off of the lighting after key presses. Therefore, it might actually be an issue with your machine / keyboard @ddB0515.
I did, however, use a virtual machine to set the static lighting mode. Thus, while compiling the openrazer debs I got an error saying that EFI variables could not be set. So it might also have to do something with that. I don't know what EFI variables openrazer tries to set and if they are even relevant for the lighting. It's just that it struck my attention, so I'm putting it here for information.
Also I love the fact that the keyboard stays in static lighting mode when disconnected from the virtual machine. It did not do that when I set the lighting mode via Synapse in a Windows VM but instead went back to the default spectrum mode as soon as I disconnected it from the VM.

**Comment by ddB0515** [comment]
Thanks @erikzmmrmnn 
I had only one more person testing it and he had same issue like I did, please check [video here](https://github.com/openrazer/openrazer/pull/1940#issuecomment-1568658199) 
And if you did same and works I would ask you to do that in normal machine (not VM) and report back
Thx

**Comment by armakunizen** [comment]
> `Bus 003 Device 006: ID 1532:02a2 Razer USA, Ltd Razer Ornata V3 X `

I compiled and installed your branch, but my device is still unrecognized. Is there a configuration file or something where I could add my USB device id on my own for locally testing?

**Comment by armakunizen** [comment]
I already changed this in `driver/razerkbd_driver.h` but maybe it's not the only place I need to edit:

```
// #define USB_DEVICE_ID_RAZER_ORNATA_V3_X 0x0294
#define USB_DEVICE_ID_RAZER_ORNATA_V3_X 0x02A2

```

**Comment by z3ntu** [comment]
@armakunizen Please see issue #2058 for your device, this one is for a different device (feel free to experiment there but let's keep this PR focused on the device here).

**Comment by armakunizen** [comment]
> @armakunizen Please see issue #2058 for your device, this one is for a different device (feel free to experiment there but let's keep this PR focused on the device here).

@z3ntu Ok, thanks for your information. I'll try this soon

**Comment by erikzmmrmnn** [comment]
@ddB0515 So I tried what you showed in the video and unfortunately it's happening to me too. I tried it in the VM as well as on the host which both run Ubuntu 22.04. What's still confusing to me is that it worked yesterday when I just selected static and did nothing else. Made me think that perhaps the issue is not with the static mode but rather another mode selected before static. I did some inspecting of this idea by resetting the keyboard via Synapse on Windows and then just selecting static in Polychromatic. However, the issue kept reappearing every time I used openrazer to set a mode so I could not prove my idea.
I also need to correct my statement from yesterday saying that the setting from Synapse was reset after disconnecting the keyboard from the Windows VM. That happened because I was using another USB controller for the Windows VM. Today when I used the same USB controller in the Windows VM that I use on the host the setting also stayed after disconnecting the keyboard from the VM and reattaching it to the host.

**Comment by ddB0515** [comment]
Last one if you could do, can you compile driver again (but remove static)
Or test that you select all other modes (except static)
As @z3ntu  and me were discussing to remove it for now (until figure it out what and why is happening ONLY on static)
If everything else works will remove static and merge it like that (for now)

Let us know here...Thanks for testing :)

**Comment by erikzmmrmnn** [comment]
I purged both openrazer and polychromatic as well as their dependencies and then reinstalled both of them again (with a reboot inbetween). After that I tried all the modes except static. However, the issue still occurs. For me it only occurs with the spectrum mode though (and static of course) but breath, reactive and wave are fine.

**Comment by gand0rf** [comment]
Hey. Just got the 1532:02A2 board. Just came across this and wanted to see how it was going and if you might need another test board.

**Comment by suchkow** [comment]
Hello! Got the same keyboard as you all (Ornata V3 X), but different PID, which is 02A2. So, it is not recognizable by any app. I see you are doing a great job rn, but in case you've missed to mention: there are two (I belive) different PIDs for Ornata V3 (maybe I am talking about two different versions "V3 X" and "V3").

**Comment by ddB0515** [comment]
Closing this as new PR is opened #2182 that handles it

**Pull request closed without merging**